drv/stm32h7-eth/src/ring.rs

```diff
@@ -0,0 +1,429 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! DMA descriptor rings and data buffers.
+//!
+//! While this module is specific to the DMA descriptor format used by the
+//! Synopsys Ethernet MAC in the STM32H7, it does _not_ depend on the hardware.
+//! This module just moves memory around very carefully.
+//!
+//! Note that the APIs on this module are entirely safe. This is deliberate. The
+//! only unsafe act that a user of this module should expect to perform is
+//! setting up the `static mut` data buffers required to call `new` on the
+//! respective ring types.
+
+use core::cell::{Cell, UnsafeCell};
+use core::sync::atomic::{AtomicU32, Ordering};
+
+/// Size of buffer used with the Ethernet DMA. This can be changed but must
+/// remain under 64kiB -- the driver initialization code refers to this constant
+/// when setting up the controller.
+pub const BUFSZ: usize = 1536;
+
+/// Opaque (to you) type alias for an Ethernet packet buffer. To use this module
+/// you need to create a static array of these somehow and provide it to `new`.
+pub struct Buffer(UnsafeCell<[u8; BUFSZ]>);
+
+/// We are careful to use `Buffer` in thread-safe ways and need it to be `Sync`
+/// so that it can be placed in a `static` by our users.
+unsafe impl Sync for Buffer {}
+
+impl Buffer {
+    /// Creates a zero-initialized buffer.
+    pub const fn new() -> Self {
+        Self(UnsafeCell::new([0; BUFSZ]))
+    }
+}
+
+/// Transmit descriptor record.
+///
+/// This is deliberately opaque to viewers outside this module, so that we can
+/// carefully control accesses to its contents.
+#[repr(transparent)]
+pub struct TxDesc {
+    tdes: [AtomicU32; 4],
+}
+
+impl TxDesc {
+    pub const fn new() -> Self {
+        Self {
+            tdes: [
```

Comment:
You could save a line or two with
```
    const ZERO: AtomicU32 = AtomicU32::new(0);
    Self {
        tdes: [ZERO; 4],
    }
```
but I'm not sure that's worth the cleverness.

Reply:
Being able to use it in two places is kind of nice (rxdesc has similar code). I'll play with it.

Reply:
Wound up adopting this, thanks!